plugins/synapticsmst/fu-plugin-synapticsmst.c

```diff
@@ -338,46 +349,33 @@ fu_plugin_update (FuPlugin *plugin,
 
 	/* Re-run device enumeration to find the new device version */
 	fu_plugin_set_status (plugin, FWUPD_STATUS_DEVICE_RESTART);
-	if (!synapticsmst_device_enumerate_device (device, error)) {
+	if (!synapticsmst_device_enumerate_device (device, data->dock_type,
+						   data->system_type, error)) {
 		return FALSE;
 	}
 	fu_device_set_version (dev, synapticsmst_device_get_version (device));
 
 	return TRUE;
 }
 
-static void
-fu_plugin_synapticsmst_redo_enumeration_cb (GUsbContext *ctx,
-					    GUsbDevice *usb_device,
-					    FuPlugin *plugin)
+void
+fu_plugin_device_registered (FuPlugin *plugin, FuDevice *device)
 {
-	guint16 pid;
-	guint16 vid;
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	const gchar *tmp;
 
-	vid = g_usb_device_get_vid (usb_device);
-	pid = g_usb_device_get_pid (usb_device);
+	/* dell plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "dell") == 0) {
+		/* only look at external devices from dell plugin */
+		if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_INTERNAL))
+			return;
 
-	/* Only look up if this was a dock connected */
-	if (vid != DOCK_NIC_VID || pid != DOCK_NIC_PID)
-		return;
+		tmp = fu_device_get_metadata (device,
+					      FU_DEVICE_METADATA_DELL_DOCK_TYPE);
 
-	/* Request daemon to redo coldplug, this wakes up Dell devices */
-	fu_plugin_recoldplug (plugin);
-}
-
-gboolean
-fu_plugin_startup (FuPlugin *plugin, GError **error)
-{
-	GUsbContext *usb_ctx = fu_plugin_get_usb_context (plugin);
-	if (usb_ctx != NULL) {
-		g_signal_connect (usb_ctx, "device-added",
-				  G_CALLBACK (fu_plugin_synapticsmst_redo_enumeration_cb),
-				  plugin);
-		g_signal_connect (usb_ctx, "device-removed",
-				  G_CALLBACK (fu_plugin_synapticsmst_redo_enumeration_cb),
-				  plugin);
+		if (tmp)
+			data->dock_type = tmp;
```

Comment:
storing this as c non-dup'd string doesn't give me the warm-and-fuzzies.

Reply:
This case I think you're right, it should be dup'ed, if the dock goes away you might get stuck with a stale pointer.